File: test-zig-persistent/module.zig
const std = @import("std");
const raft = @import("raft");

pub const std_options_debug_io: std.Io = undefined;

comptime {
    _ = raft;
}

var idx: u64 = 0;
var items: u64 = 0;
var sets: u64 = 0;

export fn _start() void {
    raft.persistent(open, update, finish, read);
    raft.streaming(streamOpen, streamRecv, streamClosed);
    raft.watchable(watchOpen, watchClosed);
}

fn open() u64 {
    return 0;
}

fn update(index: u64, cmd: []u8) raft.Result {
    items += 1;
    const value = idx;
    idx = index;
    var i: usize = 0;
    while (std.mem.indexOfPos(u8, cmd, i, "test")) |pos| {
        @memcpy(cmd[pos .. pos + 4], "best");
        i = pos + 4;
    }
    return .{ .value = value, .data = cmd };
}

fn finish() void {
    sets += 1;
}

fn read(query: []u8) raft.Result {
    if (std.mem.eql(u8, query, "index")) {
        return .{ .value = idx };
    } else if (std.mem.eql(u8, query, "items")) {
        return .{ .value = items };
    } else if (std.mem.eql(u8, query, "sets")) {
        return .{ .value = sets };
    }
    @panic("Unrecognized query");
}

fn streamOpen() void {}

fn streamRecv(data: []u8) void {
    if (std.mem.eql(u8, data, "close")) {
        raft.streamClose();
    } else {
        raft.streamSend(1, data);
    }
}

fn streamClosed() void {}

fn watchOpen(data: []u8) void {
    const n = std.fmt.parseInt(u64, data, 10) catch @panic("Invalid watch count");
    var i: u64 = 1;
    var tmp: [20]u8 = undefined;
    while (i <= n) : (i += 1) {
        const s = std.fmt.bufPrint(&tmp, "{d}", .{i}) catch unreachable;
        raft.watchSend(1, s);
    }
    raft.watchClose();
}

fn watchClosed() void {}

File: test-go-persistent/module.go
package main

import (
	"bytes"
	"strconv"

	"github.com/pantopic/ext-raft/sdk-go"
)

var (
	idx   uint64
	items uint64
	sets  uint64
)

func main() {
	raft.Persistent(open, update, finish, read)
	raft.Streaming(streamOpen, streamRecv, streamClosed)
	raft.Watchable(watchOpen, watchClosed)
}

func open() uint64 {
	return 0
}

func update(index uint64, cmd []byte) (value uint64, data []byte) {
	items++
	value = idx
	idx = index
	data = bytes.ReplaceAll(cmd, []byte(`test`), []byte(`best`))
	return
}

func finish() {
	sets++
}

var (
	queryIndex = []byte(`index`)
	queryItems = []byte(`items`)
	querySets  = []byte(`sets`)
)

func read(query []byte) (value uint64, data []byte) {
	if bytes.Equal(query, queryIndex) {
		value = idx
	} else if bytes.Equal(query, queryItems) {
		value = items
	} else if bytes.Equal(query, querySets) {
		value = sets
	} else {
		panic(`Unrecognized query: "` + string(query) + `"`)
	}
	return
}

func streamOpen() {
}

func streamRecv(data []byte) {
	if bytes.Equal(data, []byte(`close`)) {
		raft.StreamClose()
	} else {
		raft.StreamSend(1, data)
	}
}

func streamClosed() {
}

func watchOpen(data []byte) {
	n, err := strconv.Atoi(string(data))
	if err != nil {
		panic(err)
	}
	for i := range n {
		raft.WatchSend(1, []byte(strconv.Itoa(i+1)))
	}
	raft.WatchClose()
}

func watchClosed() {
}
